### PR TITLE
xt: add validation for image url

### LINKED ISSUE #4: Image URL Validation
#### As a user I want to be prevented from submitting invalid image URLs.

Acceptance criteria:
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

Discussion Topic:
- [ ] What is a valid image URL?

Dependencies:
- Save Image Link

### PR DESCRIPTION
**Pull request by Xin-Tong**
Closes #4 

The simple_form has done the front end url validation for us only if we use the `url` input tag. I add the back end validation in this PR.

### CODE REVIEW

File: app/models/image.rb [hunk 1]
```diff
@@ -1,2 +1,5 @@
 class Image < ActiveRecord::Base
+  validates_format_of :url,
+                      with: URI::regexp,
+                      message: 'must be valid'
```
Comment (bboe) [inline app/models/image.rb]: How must this field be valid? If you saw the error message "Image must be valid" what does that convey to you?

Comment (rmacklin) [inline app/models/image.rb]: The message would actually be "Url must be valid"

File: test/models/image_test.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+
+  test 'validation for nil url' do
+    image_nil_url = Image.new
+    refute_predicate image_nil_url, :valid?
```
Comment (bboe) [inline test/models/image_test.rb]: Great use of `refute_predicate`.

File: test/models/image_test.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+
+  test 'validation for nil url' do
+    image_nil_url = Image.new
```
Comment (bboe) [inline test/models/image_test.rb]: Excellent descriptive variable name.

File: test/models/image_test.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    assert_equal ['must be valid'], image_nil_url.errors[:url]
+  end
+
+  test 'validation for valid url' do
+    image_valid_url = Image.new(name: 'valid image', url: 'http://valid.com')
+    assert_predicate image_valid_url, :valid?
+  end
+
+  test 'validation for invalid url' do
+    image_invalid_url = Image.new(name: 'invalid image', url: 'invalid')
```
Comment (bboe) [inline test/models/image_test.rb]: What other invalid URL tests can you perform? Additionally what other sort of valid URL syntaxes do you want to permit?

**Comment by bboe** [comment]
This PR is a great start to this task. I see that for image URL validation you've chosen to allow all URIs. Is that restrictive enough? Ideally any URI we save should be able to render the image at a given URI. Are there URIs that the browser won't be able to automatically handle?

Similarly, do all URLs correspond to images? How could we increase the chances that the URL one is submitting actually points to an image?

Finally, there should be a flow test associated with submitting an invalid form now that validation exists. Can you add that?

File: app/models/image.rb (hunk 1 above)
Comment (bboe) [inline app/models/image.rb]: @rmacklin :) I was being imprecise as this validator may be copied for a different field, or the field name may be changed. Something like `is not a valid URL` which would be translated into `Url is not a valid URL` provides a little more context.

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -9,8 +9,11 @@ def new
 
   def create
     @image = Image.new(image_params)
-    @image.save!
-    redirect_to images_path
+    if @image.save
+      redirect_to images_path
+    else
+      render :new, status: :unprocessable_entity
```
Comment (bboe) [inline app/controllers/images_controller.rb]: :star: for using `:unprocessable_entity`

File: app/models/image.rb [hunk 6]
```diff
@@ -1,2 +1,5 @@
 class Image < ActiveRecord::Base
+  validates_format_of :url,
+                      with: /\A#{URI.regexp(%w(http https))}\z/,
+                      message:  I18n.t('errors.url_invalid')
```
Comment (bboe) [inline app/models/image.rb]: :star2: I really like that you're using `I18n` for this. Awesome!

File: app/models/image.rb [hunk 7]
```diff
@@ -1,2 +1,5 @@
 class Image < ActiveRecord::Base
+  validates_format_of :url,
+                      with: /\A#{URI.regexp(%w(http https))}\z/,
```
Comment (bboe) [inline app/models/image.rb]: Can you make this regular expression a constant in the class? You might want to use it again later.

File: test/controllers/images_controller_test.rb [hunk 8]
```diff
@@ -33,4 +33,17 @@ class ImagesControllerTest < ActionController::TestCase
...
+
+  test 'create with invalid url' do
+    image_name = 'InvalidSample'
+    image_url = 'ftp://google.com'
+    params = { image: { name: image_name, url: image_url } }
+
+    assert_no_difference 'Image.count' do
+      post :create, params
+    end
+    assert_template :new
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: Let's avoid the use of `assert_template`. `assert_select` provides what we really want to test for, i.e., that the HTTP response body contains the content we expect.

File: test/controllers/images_controller_test.rb [hunk 9]
```diff
@@ -33,4 +33,17 @@ class ImagesControllerTest < ActionController::TestCase
...
     assert_equal image_url, Image.last.url
   end
+
+  test 'create with invalid url' do
+    image_name = 'InvalidSample'
+    image_url = 'ftp://google.com'
+    params = { image: { name: image_name, url: image_url } }
+
+    assert_no_difference 'Image.count' do
+      post :create, params
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: Is there a specific reason why you made a variable for params? I personally find it more difficult to read the way you've done it, and I feel that variables should be used in order to make something _more_ readable.

**Comment by Xin-Tong** [comment]
@bboe Comments have been addressed.

File: test/flow_tests/images_flow_test.rb [hunk 10]
```diff
@@ -18,4 +18,22 @@ class ImageFlowTest < FlowTestCase
...
     assert_selector("img[src=\"#{valid_url}\"]", count: 1)
   end
+
+  test 'cannot create a image with invalid url' do
+    invalid_url = 'invalid.com'
+    visit root_path
+    click_on('Upload an image')
+    page.assert_current_path new_image_path
+
+    click_on('Main page')
```
Comment (bboe) [inline test/flow_tests/images_flow_test.rb]: Do you need to do this loop again as part of this flow? It's not really part of either of these flows to check that the navbar links work.

File: test/models/image_test.rb [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    assert_equal ['must be a valid url'], image_invalid_url.errors[:url]
+
+    image_invalid_url.url = 'ftp://google.com'
+    refute_predicate image_invalid_url, :valid?
+    assert_equal ['must be a valid url'], image_invalid_url.errors[:url]
+
+    image_invalid_url.url = 'www.image.com'
+    refute_predicate image_invalid_url, :valid?
+    assert_equal ['must be a valid url'], image_invalid_url.errors[:url]
+  end
```
Comment (bboe) [inline test/models/image_test.rb]: As we talked about, how can we dry this up? And what other URLs can we test to ensure our validation is working as we expect. What about `''` as a URL?

File: test/flow_tests/images_flow_test.rb (hunk 10 above)
Comment (Xin-Tong) [inline test/flow_tests/images_flow_test.rb]: I have addressed this issue in `ImageTag` story. This is the link to `images_flow_test.rb` of that branch: https://github.com/AppFolioOnboarding/image-sharer-xin/blob/ImageTag/test/flow_tests/images_flow_test.rb

File: test/models/image_test.rb [hunk 12]
```diff
@@ -8,21 +8,22 @@ class ImageTest < ActiveSupport::TestCase
...
-
-    image_invalid_url.url = 'ftp://google.com'
-    refute_predicate image_invalid_url, :valid?
-    assert_equal ['must be a valid url'], image_invalid_url.errors[:url]
-
-    image_invalid_url.url = 'www.image.com'
-    refute_predicate image_invalid_url, :valid?
-    assert_equal ['must be a valid url'], image_invalid_url.errors[:url]
+    image_invalid_url = Image.new(name: 'invalid image')
+    invalid_urls = %w(invalid ftp://google.com www.image.com http// http:/)
```
Comment (bboe) [inline test/models/image_test.rb]: This looks great. What about `''` and `http://`?

File: test/models/image_test.rb [hunk 13]
```diff
@@ -8,21 +8,22 @@ class ImageTest < ActiveSupport::TestCase
   end
 
   test 'validation for valid url' do
```
Comment (bboe) [inline test/models/image_test.rb]: The REGEX that you added for validation has a lot of possible allowed urls. All of the possible states the URL relative to that REGEX should be tested, including the corner cases. The outside corners (too few or too many) characters should also be tested if you're going to use such a complex REGEX.

**Comment by Xin-Tong** [comment]
@bboe Comments have been addressed on https://github.com/AppFolioOnboarding/image-sharer-xin/pull/20/commits/4d53607383f28dcd132df4a267056752ed6597d6.

I'll rebase this branch after code review.

File: test/models/image_test.rb [hunk 14]
```diff
@@ -18,7 +26,20 @@ class ImageTest < ActiveSupport::TestCase
 
   test 'validation for invalid url' do
     image_invalid_url = Image.new(name: 'invalid image')
-    invalid_urls = %w(invalid ftp://google.com www.image.com http// http:/)
+    invalid_urls = %w(invalid
+                      ''
```
Comment (bboe) [inline test/models/image_test.rb]: Does `''` here work as intended?

File: test/models/image_test.rb [hunk 15]
```diff
@@ -18,7 +26,20 @@ class ImageTest < ActiveSupport::TestCase
...
+                      ftp://google.com
+                      www.image.com
+                      http//
```
Comment (bboe) [inline test/models/image_test.rb]: In order for the "accidentally left off a single character" test to work properly, you will want to provide a almost valid URL in these cases. Such as: `http//www.google.com`.

File: test/models/image_test.rb (hunk 14 above)
Comment (Xin-Tong) [inline test/models/image_test.rb]: Thanks for catching this up. Addressed on https://github.com/AppFolioOnboarding/image-sharer-xin/pull/20/commits/174470e392d62932b081ffc66d031223c45e0cd6

File: test/models/image_test.rb (hunk 15 above)
Comment (Xin-Tong) [inline test/models/image_test.rb]: Addressed on https://github.com/AppFolioOnboarding/image-sharer-xin/pull/20/commits/174470e392d62932b081ffc66d031223c45e0cd6

File: app/models/image.rb [hunk 16]
```diff
@@ -1,3 +1,7 @@
 class Image < ActiveRecord::Base
+  VALID_URL = %r!(^(http|https)://[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.[a-z]{2,5}((:[0-9]{1,5})?/.*)?$)!
```
Comment (bboe) [inline app/models/image.rb]: There are still a few things this regex doesn't allow for that you haven't tested. For example:

```
https://a--a.com
https://a..a.com
https://a-.a.com
https://127.0.0.1
https://a.com:123456
https://#a.com
https://a.#.com
https://a.a#.com
```

Furthermore, for the valid cases that take ranges (e.g., `{2, 5}`) there should be a valid test for both:

```
https://a.co
https://a.abcde
```

The same should be done with the port range.

File: test/flow_tests/images_flow_test.rb (hunk 10 above)
Comment (bboe) [inline test/flow_tests/images_flow_test.rb]: Was this branch merged? If so, can you rebase?
